### ISSUE TITLE
 signature never resolves on session already existing. [ethereum-provider 1.5.2]

### ISSUE THREAD

**Issue by Benjythebee** [WalletConnect/walletconnect-monorepo#564]
I just did the switch from web3-provider to ethereum-provider.
It brought a few issues with it, one of which is that signing does not resolve after the first signature.

To reproduce:
Create an instance as usual and signin via QR code and call a signature; eg:
```js
await provider.enable()
signature = await provider.request({ method: 'personal_sign', params: ['hellooooo', app.provider.accounts[0]], jsonrpc: '2.0' })
```
The first time, this works totally fine, the QR code and signature fully resolves one after the other.

However, if you call this one more time, or after a refresh
```js
await provider.request({ method: 'personal_sign', params: ['hellooooo', app.provider.accounts[0]], jsonrpc: '2.0' })
```
 the signature will pop-up on your phone, but never resolve on the client side.

I'm using metamask mobile as a phone wallet
@walletConnect/ethereum-provider 1.5.2

**Edit: I have updated `@walletConnect/ethereum-provider` to 1.6.1.**

Now if I don't refresh after setting the provider and call
```js
await provider.request({ method: 'personal_sign', params: ['hellooooo', app.provider.accounts[0]], jsonrpc: '2.0' })
```
it will result in a successful signature that will resolve.

**HOWEVER** the issue is still the same after refreshing the page (even though the provider is defined after refresh).

**Comment by Shmoji** [comment]
I am having the same issue using MetaMask mobile

**Comment by Benjythebee** [comment]
bump @pedrouid ; Sorry for the tag but it seems it's a relatively widespread issue;

**Comment by tayyabmh** [comment]
I'm having this issue as well. I feel like WalletConnect is caching something and whatever that caching is happening is not bubbled up through errors.

I have to hard refresh to get something to work.

**Comment by haowang1013** [comment]
So, the root cause of the bug is in [JsonRpcProvider](https://github.com/WalletConnect/walletconnect-utils/blob/master/jsonrpc/provider/src/provider.ts#L14).

The result of the rpc call gets picked up in the onPayload callback, which gets registered in the [open](https://github.com/WalletConnect/walletconnect-utils/blob/master/jsonrpc/provider/src/provider.ts#L109) function.

And if I use the sample code earlier in this thread, the only place where open gets called is [here](https://github.com/WalletConnect/walletconnect-utils/blob/master/jsonrpc/provider/src/provider.ts#L64), and this only happens when there's no saved connection from the previous session.

This is a def a bug in the library itself rather than how it's used by the developers, and I'm a bit surprised to see it is still not fixed.

**Comment by haowang1013** [comment]
Not sure if it's the best fix but manually binding the callbacks in the [constructor](https://github.com/WalletConnect/walletconnect-utils/blob/master/jsonrpc/provider/src/provider.ts#L19) fixed it for me.

```
constructor(connection) {
    super(connection);
    this.events = new EventEmitter();
    this.connection = this.setConnection(connection);

    if (this.connection.connected) {
        this.connection.on("payload", (payload) => this.onPayload(payload));
        this.connection.on("close", () => this.events.emit("disconnect"));
        this.connection.on("error", (error) => this.events.emit("error", error));
    }
}
```

**Comment by haowang1013** [comment]
If chaning the library is not possible, a workaround can also be implemented in user space (manually calling the open function forces the event handlers to be bound)

```
async function fixWalletConnectProvider(provider) {
    // workaround for issue https://github.com/WalletConnect/walletconnect-monorepo/issues/564
    if (!provider || !provider.signer || !provider.signer.connection) {
        return
    }

    const savedConnection = provider.signer.connection
    provider.signer.connection = {
        connected: false
    }

    await provider.signer.open(savedConnection)
}

```

**Comment by Benjythebee** [comment]
I have updated walletconnect to 1.7.7 and I think this issue has been fixed as I can't reproduce both the 1.5.2 case and the 1.6.1 case.

Anyone else?

**Comment by finessevanes** [comment]
@haowang1013 @Shmoji is this still an issue?
